inject/src/test/groovy/io/micronaut/context/env/PropertySourcePropertyResolverSpec.groovy

```diff
@@ -641,4 +643,70 @@ class PropertySourcePropertyResolverSpec extends Specification {
         then:
         resolver.containsProperty("extra.listval")
     }
+
+    void "test expression resolver"() {
+        given:
+            Map<String, Object> parameters = [foo: "bar"]
+            PropertyResolver propertyResolver = new MapPropertyResolver(parameters)
+            DefaultPropertyPlaceholderResolver propertyPlaceholderResolver = new DefaultPropertyPlaceholderResolver(propertyResolver, ConversionService.SHARED);
+            propertyPlaceholderResolver.@expressionResolvers = [new PropertyExpressionResolver() {
+                @Override
+                Optional resolve(PropertyResolver pr, ConversionService conversionService, String expression, Class requiredType) {
+                    assert requiredType == String.class
+                    assert conversionService
+                    if ("foobar" == expression) {
+                        return Optional.of("ABC")
+                    }
+                    if ("xyz" == expression) {
+                        return Optional.of("123")
+                    }
+                    if ("external" == expression) {
+                        return pr.get("foo", requiredType)
+                    }
+                    return Optional.empty()
+                }
+            }]
+
+        expect:
+            def resolved = propertyPlaceholderResolver.resolvePlaceholders(template)
+            if (result) {
+                assert resolved.get() == result
+            } else {
+                assert !resolved.isPresent()
+            }
+        where:
+            template              | result
+            "Hello \${foo}!"      | "Hello bar!"
```

Comment:
Use single quote string to avoid having to escape dollars

Reply:
resolved via https://github.com/micronaut-projects/micronaut-core/pull/7422/commits/58f9d5ea7e44dce92c4da6f84549b947721b02bd